smk/binning_preparation.smk

```diff
@@ -372,44 +373,42 @@ rule filter_contigs_illumina_coas:
 ################################################################################################
 # Create BWA index for the filtered contigs
 ################################################################################################
-rule BWA_index_contigs:
-    input:
-        scaffolds="{wd}/{omics}/8-1-binning/scaffolds_{scaf_type}/batch{batch}.{min_length}.fasta"
-    output:
-        "{wd}/{omics}/6-mapping/BWA_index/scaffolds_{scaf_type}/batch{batch}.{min_length}.fasta.0123",
-        "{wd}/{omics}/6-mapping/BWA_index/scaffolds_{scaf_type}/batch{batch}.{min_length}.fasta.amb",
-        "{wd}/{omics}/6-mapping/BWA_index/scaffolds_{scaf_type}/batch{batch}.{min_length}.fasta.ann",
-        "{wd}/{omics}/6-mapping/BWA_index/scaffolds_{scaf_type}/batch{batch}.{min_length}.fasta.bwt.2bit.64",
-    shadow:
-        "minimal"
-    log:
-        "{wd}/logs/{omics}/6-mapping/scaffolds_{scaf_type}/batch{batch}.{min_length}.BWA_index.log"
-    resources:
-        mem = lambda wildcards, attempt: config['BWA_index_memory'] + 40*(attempt-1)
-    threads: 4
-    conda:
-        config["minto_dir"]+"/envs/MIntO_base.yml" #bwa-mem2
-    shell:
-        """
-        outdir=$(dirname {output[0]})
-        mkdir -p $outdir
-        prefix=$(basename {input})
-        time (
-            bwa-mem2 index {input.scaffolds} -p $prefix
-            rsync -a $prefix.* $outdir/
-        ) >& {log}
-        """
+
+# If CLUSTER_NODES is defined, then return the file symlink'ed to local drives.
+# Otherwise, return the original index files in project work_dir.
+def get_contig_bwa_index(wildcards):
+
+    # Where are the index files?
+    if CLUSTER_NODES != None:
+        index_location = 'BWA_index_local'
+    else:
+        index_location = 'BWA_index'
+
+    # Get all the index files!
+    files = expand("{wd}/{omics}/8-1-binning/scaffolds_{scaf_type}/{location}/batch{batch}.{min_length}.fasta.{ext}",
+            wd         = wildcards.wd,
+            omics      = wildcards.omics,
+            scaf_type  = wildcards.scaf_type,
+            location   = index_location,
+            batch      = wildcards.batch,
+            min_length = wildcards.min_length,
+            ext        = ['0123', 'amb', 'ann', 'bwt.2bit.64', 'pac'])
+
+    # Return them
+    return(files)
 
 # Maps reads to contig-set using bwa2
 # Baseline memory usage is:
-#   Main: BWA_memory 25GB
+#   BWA : 3.1 bytes per base in DNA database (regression: mem = 5.556e+09 + 3.011*input)
 #   Sort: using config values: SAMTOOLS_sort_threads and SAMTOOLS_sort_perthread_memgb;
+#   Total = BWA + Sort
```

Comment:
Why is this `SAMTOOLS_sort_threads` not being hard-coded like the one for abundance? Same for `SAMTOOLS_sort_perthread_memgb`: I guess it could be derived same way as for BWA-mem.

Reply:
Good point. I have now hardcoded `SAMTOOLS_sort_threads` to be 3 threads (aed33c8). 

It is more complicated to set `SAMTOOLS_sort_perthread_memgb`. We need some more work to profile many jobs and make a regression-based estimate. But we should aim to do it at some point.